runtime/parachains/src/router/hrmp.rs

```diff
@@ -62,3 +72,359 @@ pub struct HrmpChannel {
 	/// that no messages were previously added.
 	pub mqc_head: Option<Hash>,
 }
+
+/// Routines and getters related to HRMP.
+impl<T: Trait> Module<T> {
+	/// Remove all storage entries associated with the given para.
+	pub(super) fn clean_hrmp_after_outgoing(outgoing_para: ParaId) {
+		<Self as Store>::HrmpOpenChannelRequestCount::remove(&outgoing_para);
+		<Self as Store>::HrmpAcceptedChannelRequestCount::remove(&outgoing_para);
+
+		// close all channels where the outgoing para acts as the recipient.
+		for sender in <Self as Store>::HrmpIngressChannelsIndex::take(&outgoing_para) {
+			Self::close_hrmp_channel(&HrmpChannelId {
+				sender,
+				recipient: outgoing_para.clone(),
+			});
+		}
+		// close all channels where the outgoing para acts as the sender.
+		for recipient in <Self as Store>::HrmpEgressChannelsIndex::take(&outgoing_para) {
+			Self::close_hrmp_channel(&HrmpChannelId {
+				sender: outgoing_para.clone(),
+				recipient,
+			});
+		}
+	}
+
+	/// Iterate over all open channel requests and:
+	///
+	/// - prune the stale requests
+	/// - enact the confirmed requests
+	pub(super) fn process_hrmp_open_channel_requests(config: &HostConfiguration<T::BlockNumber>) {
+		let mut open_req_channels = <Self as Store>::HrmpOpenChannelRequestsList::get();
+		if open_req_channels.is_empty() {
+			return;
+		}
+
+		// iterate the vector starting from the end making our way to the beginning. This way we
+		// can leverate `swap_remove` to efficiently remove an item during iteration.
+		let mut idx = open_req_channels.len();
+		loop {
```

Comment:
Why `loop` instead of `for (idx, channel_id) in open_req_channels.iter().enumerate().rev() { let channel_id = channel_id.clone(); ... }`? Manual index handling feels like an opportunity for bugs.

Reply:
I actually tried to address this in the comment. The reason for the explicit `loop` is that we modify it while iterating on it.